File: LeetCode/Python/solutions/p1539.py
from typing import List


class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 54 ms @ (beats) 65.76%
        # Memory Usage: 14 MB @ (beats) 80.43%
        #
        # count = 0
        # n = 1
        # i = 0
        # while i < len(arr):
        #     if n != arr[i]:
        #         count += 1
        #         if count >= k:
        #             return n
        #         n += 1
        #         continue
        #     i += 1
        #     n += 1
        #
        # if count < k:
        #     return arr[-1] + k - count

        #
        # Solution v2 Binary
        #
        # Runtime: ms @ (beats) %
        # Memory Usage: MB @ (beats) %
        #
        left = 0
        length = len(arr)
        right = length - 1

        while left <= right:
            mid = left + (right - left) // 2
            missing = arr[mid] - (mid + 1)

            if missing < k:
                left = mid + 1
            else:
                right = mid - 1
                length = mid

        return length + k


if __name__ == "__main__":
    solution = Solution()

    arr = [2, 3, 4, 7, 11]
    k = 5
    print(f"return: {solution.findKthPositive(arr, k)}")  # 9

    arr = [1, 2, 3, 4]
    k = 2
    print(f"return: {solution.findKthPositive(arr, k)}")  # 6

    arr = [2]
    k = 1
    print(f"return: {solution.findKthPositive(arr, k)}")  # 1

File: LeetCode/Java/src/solutions/p1539_KthMissingPositiveNumber/P1539.java
package solutions.p1539_KthMissingPositiveNumber;

public class P1539 {
    public static void main(String[] args) {

        P1539 solution = new P1539();

        int[] arr = {2, 3, 4, 7, 11};
        int k = 5;
        System.out.println(solution.findKthPositive(arr, k));   // 9

        arr = new int[]{1, 2, 3, 4};
        k = 2;
        System.out.println(solution.findKthPositive(arr, k));   // 6

        arr = new int[]{2};
        k = 1;
        System.out.println(solution.findKthPositive(arr, k));   // 1
    }

    public int findKthPositive(int[] arr, int k) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 42.2 MB @ (beats) 29.37%
         */

        // int count = 0;
        // int n = 1;
        //
        // for (int i = 0; i < arr.length; i++) {
        //     while (n < arr[i]){
        //         count++;
        //         if (count >= k){
        //             return n;
        //         }
        //         n++;
        //     }
        //     n++;
        // }
        //
        // int last = arr[arr.length - 1];
        // while (count < k) {
        //     last++;
        //     count++;
        // }
        // return last;

        /*
         * Solution v2: Binary
         *
         * Runtime:  ms @ (beats) %
         * Memory Usage:  MB @ (beats) %
         */
        int n = arr.length;
        int l = 0;
        int r = n - 1;

        while (l <= r) {
            int mid = l + (r - l) / 2;
            int missing = arr[mid] - (mid + 1);
            if (missing < k) {
                l = mid + 1;
            } else {
                r = mid - 1;
                n = mid;
            }
        }
        return n + k;

    }
}
